dbms/src/Storages/Page/V3/PageEntryCheckpointInfo.h

```diff
@@ -0,0 +1,83 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Storages/Page/V3/CheckpointFile/Proto/manifest_file.pb.h>
+
+namespace DB::PS::V3
+{
+
+namespace CheckpointProto
+{
+using StringsInternMap = std::unordered_map<std::string_view, std::shared_ptr<const std::string>>;
+}
+
+struct CheckpointLocation
+{
+    // This struct is highly coupled with manifest_file.proto -> EditEntry.
+
+    std::shared_ptr<const std::string> data_file_id;
+
+    uint64_t offset_in_file = 0;
+    uint64_t size_in_file = 0;
+
+    CheckpointProto::EntryDataLocation toProto() const
+    {
+        CheckpointProto::EntryDataLocation proto_rec;
+        proto_rec.set_data_file_id(*data_file_id);
```

Comment:
There will be many duplicate `data_file_id` and consume lots of disk space. Can we use a dictionary to optimize?

Reply:
The manifest file is written over a LZ4 compressed format. Duplicated data file ids will be eliminated away.

I tested with 10000 records with the same 76 byte long DataFileId. The generated manifest file is 92KB which means the materialized size is only 9.2 bytes / record.

Reply:
The compression ratio looks good.

Since we need to put all those file_ids into a temporary object `ManifestFileEditsPart`, and the file_ids are already normalized, another thing is the memory consumption could be large. E.g. 10 000 000 records with 76 bytes each takes 724MiB. It looks acceptable.

@lidezhu Maybe we can also using LZ4 like here to generate GC dump file during PageStorage Full GC.

Reply:
@breezewish BTW, is it possible to use a streaming way to generate `ManifestFileEditsPart` in the future? I'm concerned that the page number could be larger in extreme cases.

Reply:
> Since we need to put all those file_ids into a temporary object ManifestFileEditsPart, and the file_ids are already normalized, another thing is the memory consumption could be large. E.g. 10 000 000 records with 76 bytes each takes 724MiB.

It's true. Our API allows caller to split into multiple calls to reduce temp memory overhead as well. For example, we may split into multiple writes with each write consumes 10000 records from the memory. This effectively reduces peak temp memory overhead to only 760KB.